README.md
# peferb/trafiklab 
#### _github.com/peferb/trafiklab@v0.2.4 ([changelog](https://github.com/peferb/trafiklab/blob/main/CHANGELOG.md))_

Golang wrapper of [Trafiklab.se](https://trafiklab.se/) which supplies traffic information about Swedish trains, buses, trams and ships.

Implemented to match the official specifications as much as possible while trying to keep it simple and small and also
easy to use. 

**GO GET IT!**
```shell 
$ go get github.com/peferb/trafiklab@v0.2.4
```

## CLI example

```go
package main

import (
	"bytes"
	"encoding/json"
	"github.com/peferb/trafiklab/sl"
	. "github.com/peferb/trafiklab/sl/api"
	"log"
	"net/url"
	"os"
)

func main() {
	departuresApi := sl.NewApi(RealTimeDeparturesV4, JSON, "you api key here")
	departures, err := departuresApi.GetBytes(url.Values{"siteid": {os.Args[1]}})
	if err != nil {
		log.Println(err)
		return
	}

	prettyJson := &bytes.Buffer{}
	err = json.Indent(prettyJson, departures, "", "  ")
	log.Println(prettyJson.String(), err)
	// Example run:
	// $ go run . 9710
}
```

## Server example
```go
package main

import (
	"github.com/peferb/trafiklab/sl"
	. "github.com/peferb/trafiklab/sl/api"
	"log"
	"net/http"
)

func departuresHandler(w http.ResponseWriter, req *http.Request) {
	departuresApi := sl.NewApi(RealTimeDeparturesV4, JSON, "your api key here")
	
	bytes, err := departuresApi.GetBytes(req.URL.Query())
	if err != nil {
		w.WriteHeader(500)
		w.Write([]byte(err.Error()))
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(200)
	w.Write(bytes)
}

func main() {
	http.HandleFunc("/realtime-departures", departuresHandler)
	log.Println("Listening for requests at http://localhost:8000/")
	log.Fatal(http.ListenAndServe(":8000", nil))
	// Example url: 
	// localhost:8000/realtime-departures?siteid=9710
}
```

## Setup

* Sign up at [trafiklab.se](https://trafiklab.se)
* Generate keys

### SL Departures v4.0 (Realtidsinformation 4)
* [Generate API key](https://developer.trafiklab.se/api/sl-realtidsinformation-4)
* [API Documentation](https://www.trafiklab.se/api/trafiklab-apis/sl/departures-4/)

### SL Service alerts v2.0 AKA Deviations (Störningsinformation 2)
* [Generate API key](https://developer.trafiklab.se/api/sl-storningsinformation-2)
* [API Documentation](https://www.trafiklab.se/api/trafiklab-apis/sl/service-alerts-2/)

### SL Stop lookup v1.0 AKA TypeAhead (Platsuppslag)
* [Generate API key](https://developer.trafiklab.se/api/sl-platsuppslag)
* [API Documentation](https://www.trafiklab.se/api/trafiklab-apis/sl/stop-lookup/)

### SL Traffic status v2.0 AKA Traffic Situation (SL Trafikläget 2) 
* [Generate API key](https://developer.trafiklab.se/api/sl-trafiklaget-2)
* [API Documentation](https://www.trafiklab.se/api/trafiklab-apis/sl/sl-traffic-status-2/)

### SL Stops and lines v2.0 AKA LineData (SL Hållplatser och Linjer 2)
* [Generate API key](https://developer.trafiklab.se/api/sl-hallplatser-och-linjer-2)
* [API Documentation](https://www.trafiklab.se/api/trafiklab-apis/sl/stops-and-lines-2/)

### SL Nearby stops v2.0 (Närliggande hållplatser 2)
* [Generate API key](https://developer.trafiklab.se/api/sl-narliggande-hallplatser-2)
* [API Documentation](https://www.trafiklab.se/api/trafiklab-apis/sl/nearby-stops-2/)

### SL Route Planner V3.1 AKA Travel Planner V3.1 (SL Reseplanerare 3.1)
⚠️ _"Route Planner V3.1" differs a bit from the other API:s by having multiple endpoints, and two of the endpoints only 
return data in XML format._
* [Generate API key](https://developer.trafiklab.se/api/sl-reseplanerare-31)
* [API Documentation](https://www.trafiklab.se/api/trafiklab-apis/sl/route-planner-31/)


CHANGELOG.md
# peferb/trafiklab changelog

## v0.2.4
### Route Planner V3.1 AKA Travel Planner V3.1 (SL Reseplanerare 3.1)

Added support for [Route Planner V3.1](https://www.trafiklab.se/api/trafiklab-apis/sl/route-planner-31/). But it's not a 
pretty implementation because Route Planner API breaks the general SL API structure.

#### How does Route Planner V3.1 diff against the other API:s?
* All other API:s consists of one endpoint per API. Route Planner V3.1 consists of multiple endpoints. 
* All other API:s offers JSON as data return format. In Route Planner two of the endpoints only returns XML, I have 
suffixed those with "Xml".

## v.0.2.3
### Changed all parameters to stlib types
Now DataFormat type and "Api" Type is replaced by regular strings.

I came to his decision after reading [Jack Lindamoods article "Aspects of a good Go library"](https://medium.com/@cep21/aspects-of-a-good-go-library-7082beabb403).

### NewApi returns Api...
... and not Struct{Api} as before. Don't even know what Struct{Api} is. The previous return type was the result of a 
late late night in a new lang 😅

## v0.2.1 / v0.2.2
_(duplicated version because of me messing things up)_

### Breaking changes
* Replaced the package import usage with `sl.NewApi()`.
* Moved data structure selection from func name to param. 

#### Example new usage 
```go
import (
	"github.com/peferb/trafiklab/sl"
	. "github.com/peferb/trafiklab/sl/api"
)

func main() {
	departuresApi := sl.NewApi(RealTimeDeparturesV4, JSON, "you api key here")
}
```

### New functionality
* Added "SL Nearby stops v2.0 (Närliggande hållplatser 2)"
* Added "SL Stops and lines v2.0 AKA LineData (SL Hållplatser och Linjer 2)"

### Misc
* Much smaller codebase / lib size
* Less bloated dev interface
* Added tests

## v0.1.0

#### Includes packages for:
* SL Departures v4.0 (Realtidsinformation 4)
* SL Service alerts v2.0 AKA Deviations (Störningsinformation 2)
* SL Stop lookup v1.0 AKA TypeAhead (Platsuppslag)
* SL Traffic status v2.0 AKA Traffic Situation (SL Trafikläget 2)

#### Example usage
```go
import (
	realtime "github.com/peferb/trafiklab/sl/realtime_departures_v4"
)

func main() {
	bytes, err := realtime.GetBytesJson("your api key here", req.URL.Query())
}
```

sl/api/api.go
package api

import (
	"net/http"
	"net/url"
)

type Api struct {
	GetBytes    func(query url.Values) ([]byte, error)
	GetString   func(query url.Values) (string, error)
	GetResponse func(query url.Values) (*http.Response, error)
}

// JSON Dataformat
const JSON = "json"

// XML Dataformat
const XML = "xml"

// RoutePlannerV3_1TtiXml Returns available dates for searches in "Route Planner V3.1 AKA Travel Planner V3.1 (SL Reseplanerare 3.1)"
//
// -> ALWAYS RETURNS XML <-
//
// Access
// https://developer.trafiklab.se/api/sl-reseplanerare-31
//
// Documentation
// https://www.trafiklab.se/api/trafiklab-apis/sl/route-planner-31/
const RoutePlannerV3_1TtiXml = "api.sl.se/api2/TravelplannerV3_1/tti"

// RoutePlannerV3_1XsdXml Returns all response types of "Route Planner V3.1 AKA Travel Planner V3.1 (SL Reseplanerare 3.1)"
//
// -> ALWAYS RETURNS XML <-
//
// Access
// https://developer.trafiklab.se/api/sl-reseplanerare-31
//
// Documentation
// https://www.trafiklab.se/api/trafiklab-apis/sl/route-planner-31/
const RoutePlannerV3_1XsdXml = "api.sl.se/api2/TravelplannerV3_1/xsd"

// RoutePlannerV3_1Trip Trip part of "Route Planner V3.1 AKA Travel Planner V3.1 (SL Reseplanerare 3.1)"
//
// Access
// https://developer.trafiklab.se/api/sl-reseplanerare-31
//
// Documentation
// https://www.trafiklab.se/api/trafiklab-apis/sl/route-planner-31/
const RoutePlannerV3_1Trip = "api.sl.se/api2/TravelplannerV3_1/trip"

// RoutePlannerV3_1Reconstruction Reconstruction part of "Route Planner V3.1 AKA Travel Planner V3.1 (SL Reseplanerare 3.1)"
//
// Access
// https://developer.trafiklab.se/api/sl-reseplanerare-31
//
// Documentation
// https://www.trafiklab.se/api/trafiklab-apis/sl/route-planner-31/
const RoutePlannerV3_1Reconstruction = "api.sl.se/api2/TravelplannerV3_1/reconstruction"

// RoutePlannerV3_1GisRoute Gis Route part of "Route Planner V3.1 AKA Travel Planner V3.1 (SL Reseplanerare 3.1)"
//
// Access
// https://developer.trafiklab.se/api/sl-reseplanerare-31
//
// Documentation
// https://www.trafiklab.se/api/trafiklab-apis/sl/route-planner-31/
const RoutePlannerV3_1GisRoute = "api.sl.se/api2/TravelplannerV3_1/gisroute"

// http://api.sl.se/api2/TravelplannerV3_1/gisroute.&lt;FORMAT&gt;?key=&lt;DIN APINYCKEL&gt;&amp;ctx=

// RoutePlannerV3_1JourneyDetail Journey Detail part of"Route Planner V3.1 AKA Travel Planner V3.1 (SL Reseplanerare 3.1)"
//
// Access
// https://developer.trafiklab.se/api/sl-reseplanerare-31
//
// Documentation
// https://www.trafiklab.se/api/trafiklab-apis/sl/route-planner-31/
const RoutePlannerV3_1JourneyDetail = "api.sl.se/api2/TravelplannerV3_1/journeydetail"

// RealTimeDeparturesV4 "Realtime Departures V4 (Realtidsinformation 4)"
//
// Access
// https://developer.trafiklab.se/api/sl-realtidsinformation-4
//
// Documentation
// https://www.trafiklab.se/api/trafiklab-apis/sl/departures-4/
const RealTimeDeparturesV4 = "api.sl.se/api2/realtimedeparturesV4"

// ServiceAlertsV2 "SL Stop lookup v1.0 AKA deviations (Platsuppslag)"
//
// Access
// https://developer.trafiklab.se/api/sl-platsuppslag
//
// Documentation
// https://www.trafiklab.se/api/trafiklab-apis/sl/stop-lookup/
const ServiceAlertsV2 = "api.sl.se/api2/deviations"

// StopLookupV1 "SL Stop lookup v1.0 AKA TypeAhead (Platsuppslag)"
//
// Access
// https://developer.trafiklab.se/api/sl-platsuppslag
//
// Documentation
// https://www.trafiklab.se/api/trafiklab-apis/sl/stop-lookup/
const StopLookupV1 = "api.sl.se/api2/typeahead"

// TrafficStatusV2 "SL Traffic status v2.0 AKA Traffic Situation (SL Trafikläget 2)"
//
// Access
// https://developer.trafiklab.se/api/sl-trafiklaget-2
//
// Documentation
// https://www.trafiklab.se/api/trafiklab-apis/sl/sl-traffic-status-2/
const TrafficStatusV2 = "api.sl.se/api2/trafficsituation"

// StopsAndLinesV2 "SL Stops and lines v2.0 AKA LineData (SL Hållplatser och Linjer 2)"
//
// Access
// https://developer.trafiklab.se/api/sl-hallplatser-och-linjer-2
//
// Documentation
// https://www.trafiklab.se/api/trafiklab-apis/sl/stops-and-lines-2/
const StopsAndLinesV2 = "api.sl.se/api2/LineData"

// NearbyStopsV2 "SL Nearby stops v2.0 (Närliggande hållplatser 2)"
//
// Access
// https://developer.trafiklab.se/api/sl-narliggande-hallplatser-2
//
// Documentation
// https://www.trafiklab.se/api/trafiklab-apis/sl/nearby-stops-2/
const NearbyStopsV2 = "api.sl.se/api2/nearbystopsv2"
